### PR TITLE
Concentrated barbers aid can give hairless species hair + fixes hair updating once and for all (probably)

### LINKED ISSUE #68563: Weird behavior with the Smooth-Headed quirk and HARS mutation
Reporting client version: 514.1566

## Round ID:
[186875](https://scrubby.melonmesa.com/round/186875)

## Testmerges:
- [[Ready for Review] Admin lua scripting](https://www.github.com/tgstation/tgstation/pull/65635)
- [Reworks pAIs](https://www.github.com/tgstation/tgstation/pull/68241)
- [New Satchel and Backpack Sprites!](https://www.github.com/tgstation/tgstation/pull/68524)

## Reproduction:

During a Radiation Storm, i gained the HARS mutation. Despite this, i had the moodlet for being upset about not wearing headware. Upon having the mutation removed by mutadone, i had my natural hair restored (which i did not have before HARS). I still, though, had the moodlet for wanting to cover my bald head. Wearing my wig worked as normal in this state
Step 1: Have the smooth head quirk
Step 2: Gain hars
Step 3: cry because you have low mood
Step 4: lose hars
Step 5: Stare in bewilderment at your head full of hair, but still cry until you find a hat or your wig.
image
^Me, gaining HARS during a radstorm and checking mood

### PR DESCRIPTION
**Pull request by Y0SH1M4S73R**
## About The Pull Request

Concentrated barbers aid will give hair to hairless species after a number of cycles inversely proportional to its purity. At 100% purity, it will also remove the baldness quirk. Also fixes what I assume to be every single bug related to hair not updating properly.

## Why It's Good For The Game

Hairless species can give themselves hair already by overdosing on spray tan and reapplying a backup of their genome. Most of the people that use this characteristic of spray tan find the other effects of the overdose obnoxious. I also suspect that most uses of spray tan overdose nowadays are to gain hair as a hairless species. This provides a way to bypass most of the hassle of the spray tan approach.

Regarding the fix, please point out any hair-related issues this fixes. (Off the top of my head, it fixes #68563)

## Changelog

:cl:
add: Concentrated barbers aid will give hairless species hair after it has remained in them for a number of cycles inversely proportional to the purity (20 cycles at 100% purity). Additionally, at 100% purity, it removes the smooth-headed quirk at the same time that it grants hairless species hair.
fix: Most effects that change your hairstyle are no longer delayed until certain effects occur.
fix: You cannot bypass the smooth-headed quirk with the genetics console (even if you were somehow able to figure out how to get a specific hairstyle)
fix: The species and antagonist character icons in the preferences menu are no longer all bald.
/:cl:

### CODE REVIEW

**Comment by necromanceranne** [comment]
Even rockmens might know the beauty of a pompadour.

**Comment by wesoda25** [comment]
Every PR that enables hair lizards moves us a step further from God.

File: code/modules/reagents/chemistry/reagents/other_reagents.dm [hunk 1]
```diff
@@ -2228,7 +2228,29 @@
...
+	if(current_cycle > 20 / creation_purity)
+		if(!ishuman(M))
+			return
+		var/mob/living/carbon/human/human_mob = M
+		if(creation_purity == 1 && human_mob.has_quirk(/datum/quirk/item_quirk/bald))
+			human_mob.remove_quirk(/datum/quirk/item_quirk/bald)
+		var/datum/species/species_datum = human_mob.dna?.species
+		if(!istype(species_datum))
+			return
+		if(species_datum.species_traits.Find(HAIR))
```
Comment (Kapu1178) [inline code/modules/reagents/chemistry/reagents/other_reagents.dm:2244]:
```suggestion
		if(hair in species_datum.species_traits))
```

File: code/modules/reagents/chemistry/reagents/other_reagents.dm [hunk 2]
```diff
@@ -2228,7 +2228,29 @@
...
+/datum/reagent/concentrated_barbers_aid/on_mob_life(mob/living/carbon/M, delta_time, times_fired)
+	. = ..()
+	if(current_cycle > 20 / creation_purity)
+		if(!ishuman(M))
+			return
+		var/mob/living/carbon/human/human_mob = M
+		if(creation_purity == 1 && human_mob.has_quirk(/datum/quirk/item_quirk/bald))
+			human_mob.remove_quirk(/datum/quirk/item_quirk/bald)
+		var/datum/species/species_datum = human_mob.dna?.species
+		if(!istype(species_datum))
```
Comment (Kapu1178) [inline code/modules/reagents/chemistry/reagents/other_reagents.dm:2242]:
```suggestion
		if(!species_datum)
```

**Review by Kapu1178 — CHANGES_REQUESTED** [review]

File: code/modules/reagents/chemistry/reagents/other_reagents.dm (hunk 1 above)
Comment (Mothblocks) [reply code/modules/reagents/chemistry/reagents/other_reagents.dm:2244]: Do you mean `HAIR`?

Comment (Y0SH1M4S73R) [reply code/modules/reagents/chemistry/reagents/other_reagents.dm:2244]: I thought the `list.Find(thing)` syntax was faster than the `thing in list` syntax.

**Comment by LemonInTheDark** [comment]
Don't like lizards with hair
The quirk thing seems just wrong. but I may just have a stick up my ass

File: code/modules/reagents/chemistry/reagents/other_reagents.dm (hunk 1 above)
Comment (Mothblocks) [reply code/modules/reagents/chemistry/reagents/other_reagents.dm:2244]: Not sure but this code is really cold so it doesn't matter anyway

Choose whatever you think looks best, though we use `in` most of the time

**Comment by Farquaar** [comment]
> 
> We generally try to avoid ways to remove quirks during rounds to not allow people to game the points system but since smooth headed is 0 points this is not really an issue

I suppose that makes sense.

**Comment by timothymtorres** [comment]
Looks like #68828 and #68818 might be from this PR.

**Pull request merged**